Iph improvements

emcangi opened the pull request:
## Pull Request Checklist
 - [x] Code Follows PEP8 guidelines (explain any exceptions)
 - [x] New routines don't duplicate existing functionality
 - [x] Original authors of modified routines have been added as reviewers
 
Initial PR addressing Issue #78.

FYI only the most recent commit is actually important. Unclear why the earlier 4 are even on this branch.

emcangi committed:
Add a test file and push it - just want to make sure I can push from my new computer.

emcangi committed:
Delete the file that I just tested.

emcangi committed:
Merge remote-tracking branch 'upstream/master'

Merging PR #75 commit into my working copy

emcangi committed:
Merge remote-tracking branch 'upstream/master', incorporating IPH updates.

emcangi committed:
Add a support function predict_IPH_linecenter(), which uses spice kernels to predict where the IPH line center will be. Because this can be done for every integration at once, this requires some structural changes to line_fit_initial_guess.

planetarymike committed:
Added logic to fit an IPH component based on whether the MRH altitude is greater than 100 km.

planetarymike committed:
comment out unnecessary print statement for hessian inverse

emcangi committed:
Resolve merge conflicts in echelle_graphics.py; add new code organization, printing of min chi sq, IPH component fitting, and some bug fixes (nansum on uncertainties, removal of extra keyword args, etc).

emcangi committed:
Fix docstring to include IPH parameters.

emcangi committed:
Remove IPH_bounds argument and code that plots the IPH initial guess area, since we now predict it.
Fix printing of vertical rule line labels so that IPH and H labels don't overlap if the emission lines overlap.
Add a nonzero bound to the residuals axis in the event that the fit fails (because in that scenario, a figure is still returned but with nothing on it).

emcangi committed:
Fix a bug that was causing an exception when using the dynesty fitter (modeled_params/fit_uncert and _fit_parameter_IPH_idxs are both lists, so can't index one by the other).

emcangi committed:
Add IPH width constrained by data/modeling, rather than arbitrary values.
Delete some old comment lines and whitespace.
Update lower bound for IPH brightness to be 0 for both fitting routines.

emcangi committed:
Unify code that manages model parameter names, lists of them, and dictionaries with them.
Rename all dictionary key names to the new scheme.
Create a list of uncertainty parameter names similar to _fit_parameter_names.
Update docstrings.

emcangi committed:
Revert a change in cd776d30709bc0c6734399496b6ecb5191f95f89 to the IPH brightness lower bound with the Dynesty solver, and update the lower bound for IPH brightness with Scipy fitter to be None (allows negative values in both solvers).

emcangi committed:
Fix call to line_fit_initial_guess in quicklook routine to reflect recent changes.
Create a variable for the number of model parameters so it's exceedingly clear what the 8 means, and it will be clear where to change it in the future if need be.

emcangi committed:
Add in initial guess for IPH width that is actually within the bounds.

emcangi committed:
Update quicklook code, which includes a fit to coadded data, to work with updates to line_fit_initial_guess.

emcangi commented:
As of this morning, remaining task is to get the IPH brightness onto the quicklooks.

emcangi committed:
Factor out the code that checks if IPH is fittable so it can be called elsewhere.
Update the quicklook fit process to include a print out of the IPH brightness and uncertainty--with colored text depending if blueshifted or redshifted!
Update the quicklook fit process to use existing routines (plot_line_fit, convert_to_physical_units) rather than repeating the code within them.
Move color for wavelength guidelines to top level for use in more than one function.
Correct bug in call to make_fit_param_dict in quicklook code
Adjust the alignment of a few objects on the quicklook plot (text if model fit fails, etc)
Delete line in echelle_graphics.py that fills the D central wavelength in the fit parameter dictionary since this is now handled elsewhere.

emcangi commented:
IPH brightness now added to QL in 49e215541c43a47085199ea2bb3b2bc9ad5a36d4. Ready for a final review

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -1607,34 +1634,41 @@ def fit_flat_data(light_fits, spectrum, data_unc, bad_frames=None,
-        result_vec = fit_H_and_D(initial_guess, wavelengths, spectrum[i, :], light_fits, theCLSF, unc=data_unc[i, :], BU_bg=BU_bg_i, **kwargs)
-            
+        # Determine whether line-of-sight minimum ray height is large enough to fit an IPH component
+        fit_IPH_component = check_whether_IPH_fittable(light_fits)
```
technically this is duplicating work because the full FITS file is checked on every integration loop, but eh

emcangi replied:
Fixed in 1dc86bfe73b86b1aad026fa19d45487656e827a4

planetarymike replied:
I think the new implementation is still redoing all of the work in every loop, but this is probably not an important optimization to make.

emcangi replied:
Yes this seems to have been an oversight in trying to adapt to quicklooks. it has been fixed for real this time in 9756bee44f864647d52fa08bab306de82ab6bdc3

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -1542,17 +1546,40 @@ def flatten(light_fits, processed_data):
+                        'width_IPH',  # nm
+                        'background_m',
+                        'background_b']
+_unc_parameter_names = ['unc_total_brightness_H', 
```
should probably use a list comprehension here for safety: `['unc_'+pname for pname in _fit_parameter_names]`

emcangi replied:
Fixed in 38c81f7ccf7737d2e3e3e95e56dc52065f78456a

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -1892,10 +1908,10 @@ def writeout_l1c(light_l1a_path, dark_l1a_path, l1c_savepath, light_fits, binnin
+    H_brightnesses = [fit_params_list[i]['total_brightness_H'] for i in range(n_int)]
+    D_brightnesses = [fit_params_list[i]['total_brightness_D'] for i in range(n_int)]
+    H_1sig = [fit_unc_list[i]['unc_total_brightness_H'] for i in range(n_int)]
+    D_1sig = [fit_unc_list[i]['unc_total_brightness_D'] for i in range(n_int)]
```
we should probably report the fitted IPH parameters in the l1c file, right?

emcangi replied:
Oh nooooo, yes

planetarymike replied:
Hopefully it's not too much extra work!

emcangi replied:
Fixed with c9545179fb52629ce74c4971ee40e51fe6117a2b

emcangi replied:
Fixed in c9545179fb52629ce74c4971ee40e51fe6117a2b

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2070,24 +2120,34 @@ def fit_H_and_D(pig, wavelengths, spec, light_fits, CLSF, unc=1, IPH_bounds=(Non
+                                               (None, None),  # DN_D
+                                               (None, None), # DN_IPH
+                                               (121.55, 121.58),  # λ H
+                                               (pig[4]-IPH_wv_spread/2, 
```
I think +/- 5 km/s may be safer, this is +/- 2.5 km/s, right?

emcangi replied:
Ah ok, I misunderstood and did 5 as the total spread, apparently I can't read

emcangi replied:
Fixed in 2b5aa95a4a3260adfed691da8f079d988d183a70

planetarymike replied:
I think you either need to update the comment in constants to refer to 10 km/s or change the constant back and remove the /2 from the bounds array.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2298,55 +2365,54 @@ def lineshape_model(params, wavelength_data, binedges, theCLSF, BU_bg):
-                                         # the starting guess/requirement.
-    width_IPH = params[5]  # TODO: Should also be able to calculate this, 
-                            # which is the width of the Gaussian. 
+    param_dict = make_fit_param_dict(params, BU_bg=BU_bg)
```
nice, this is more elegant. Changing this function to a class with static class members for the names / bounds / etc. is probably the move to avoid having things at the global scope, but the current implementation is good enough.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2630,43 +2697,102 @@ def line_fit_initial_guess(wavelengths, spectrum, H_a=20, H_b=170, D_a=80, D_b=1
+    guesses = np.zeros((rows, num_params))
+    # Line center of IPH is predicted based on the obs geometry;
+    # length is number of integrations.
+    lambda_IPH_lya_guess = 121.567 + predict_IPH_linecenter(light_fits)
```
probably should fit the IPH parameter the same way as D, by fitting the H central wavelength and then subtracting the IPH line center offset from the lab wavelength, but if this works that's fine for now.

planetarymike replied:
let me know if you're planning on making this change or just leaving it as is.

emcangi replied:
If we do this, then what do we do with the IPH velocity uncertainty? Currently that's ±5 km/s and that provides the bounds over which to search for the line center.

planetarymike replied:
We keep it because it represents the uncertainty in the offset relative to the mars rest wavelength. The issue is whether a "0" fit parameter for the IPH offset within the bounds of +/-5 km/s represents zero offset from the lab frame wavelength of Lyman alpha or from the fitted instrument Mars H Ly a wavelength. I think it should be the latter to minimize potential systematics involved in referencing the IPH to the lab wavelength but fitting the observed location of the Mars H line center.

emcangi replied:
After discussion, I implemented this in 6795164877bbaa7fd02e2528f26cfc204b03e771, but found that it adds a huge amount of overhead to the fitting for some reason - 5-10 seconds per call, which is not sustainable. Thus I've reverted it in 4fa5e9b4d53c3d383802df8f5c6abeb51708496c and for now we'll just accept that there will be a slight discrepancy that should be accounted for within the uncertainties anyway.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2630,43 +2697,102 @@ def line_fit_initial_guess(wavelengths, spectrum, H_a=20, H_b=170, D_a=80, D_b=1
+    Exact location of upstream direction not completely certain. 
+    Not accounted for: decelreation of IPH across solar system, assumed to be the same.
+    """
+    load_iuvs_spice()
```
I think this will clear and load the spice kernels every time this function is called... possibly greatly impacting performance. May be better to check whether the kernels are loaded and return an error if they are not, or put the contents of the function into a try/except block that asks the user to load_iuvs_spice() if a spice error is thrown.

emcangi replied:
Could you possibly add a change for this? I'm still working on error control for problem files.

emcangi replied:
We discussed in person and found that on my machine, this takes ~microseconds, and so determined not to fix this for now. Let's keep an eye on it in case it becomes a problem for us or future users.

planetarymike reviewed (commented):
I did not review changes to the graphics routines.

This looks ready to merge depending on your thoughts about some of my suggestions. I think the main thing that's still missing is ensuring the IPH fit values are reported in the l1c files. We'll need to make some minor changes to the SIS I guess once these are slated for delivery.

emcangi committed:
Avoid repeating work.

emcangi committed:
Simplify list comprehension

emcangi committed:
Correct wavelength spread to actually ±5, not 5 = total spread

emcangi committed:
Add IPH to l1c writeout. Note: This will not actually write to the l1c unless you have the appropriate lines added in IUVSFieldDesc.txt.

emcangi committed:
Allow re-calculation of the approximate Hessian with a bigger step size (epsilon) in the event where nans are generated when the IPH is fit. This new epsilon (1e-2) seems to generally allow real values to come through, but the frequency of nans appearing and how frequently this fixes it has not yet been characterized
Add the ability to change bound and bootstrap when calling dynesty.
Slightly tighten the tolerances on the call to scipy.optimize.minimize
Slightly restrict the priors for total brightness of H and D, as well as background parameters, to improve speed of dynesty call.

planetarymike commented on maven_iuvs/echelle.py:
```diff
@@ -2160,11 +2168,19 @@ def fit_H_and_D(pig, wavelengths, spec, light_fits, CLSF, unc=1, IPH_bounds=(Non
+                # automatically. Check and recalculate if need be.
+                if np.isnan(fit_uncert).any():
+                    new_hessian = approx_hess2(bestfit.x, negloglikelihood, 
+                                   epsilon=1e-2, # Seems to be the magic number...
```
Based on the approx_hess2 documentation (available here: https://www.statsmodels.org/dev/generated/statsmodels.tools.numdiff.approx_hess2.html ), the default is to use

`epsilon = EPS**(1/3)*x`

That is, the epsilon used in the hessian computation is a fraction of the fitted parameter value. The default value for EPS is `np.finfo(float).eps`, which on my machine is ~2e-16, giving `epsilon ≈ 6e-6*x` .

Probably we should also scale epsilon by the best fit value, so I recommend trying `epsilon=1e-2*bestfit.x` .

emcangi replied:
This is no longer relevant thanks to the jax implementation, thank you!

emcangi committed:
Remove the call to binning_df, which is the output of get_binning_df and includes hard-coded values that can be queried directly from the FITS files. Not yet removing get_binning_df() in case of unhandled cases.

emcangi committed:
Loosen restrictions for finding nan in uncertainties; allow nan for uncertainty on the background fit.

emcangi committed:
Revert to using idl_pipeline_folder since we haven't committed the writeout script to the python repo.

(cherry picked from commit 34d3c60bc9f19fa4b1018bb7d517b1102899c1cb)

emcangi committed:
Modify the creation of parameter dictionaries in the event of a failed fit so that they use the existing parameter name lists rather than re-hard-coding them.

emcangi committed:
Allow skipping files if they already exist.

emcangi committed:
Ensure the selection of files for lights only looks at light files.

planetarymike committed:
Transitioned echelle lineshape_model and (neg)loglikelihood function to use jax just-in-time compilation

planetarymike commented:
Pushed some changes to switch the lineshape_model and (neg)loglikelihood to use JAX. This has two advantages:

1. The (neg)loglikelihood function is just-in-time compiled, which reduces function runtime. Fitting with dynesty now takes ~15 s (static sampler) or ~30s (dynamic sampler).
2. The hessian can be computed via automatic differentiation rather than finite differences. Hopefully this results in fewer singular hessian matrices.

I have not tested this code for all modes (in particular, I have not tested the BU_background specification, which needed to be altered for the code to work with JAX). It's possible modes other than the standard mode are broken.

JAX is a bit tricky to work with if conditional evaluation is needed, but it's not too bad, hopefully this doesn't add a ton of overhead when future changes are needed to the lineshape model.

emcangi commented:
Noting that we tested this together and it looks like the dynesty fitter with the following options works at a ~1s/integration on my machine:
```
convert_l1a_to_l1c(light_fits, dark_fits, light_path, dark_path, l1c_savepath, 
                   fitter="dynesty", approach="static", bound="single", livepts=100)   # <-------------------
```

That's still slower than scipy, I think, but without the problems associated with approx_hess2, so I think it's worth it.

emcangi committed:
Split up the logic that gets the mean minimum ray height and then selects an integration from it to determine if IPH can be fit. It is modularized to support differing use in fitting individual integrations and the coadded integration (on quicklooks).

emcangi committed:
Update the default values for the fitting routines to use dynesty's static sampler by default, along with the most efficient choices for the dynesty solver arguments, determined empirically (e.g. by fussing around until satisfied).
Update the priors for the IPH to allow a greater range (previously it was failing on observations with IPH and H Ly alpha that are very similar brightness).

emcangi committed:
Move get_mean_mrh() to geometry as it fits better there.

emcangi committed:
Add support for processing nonlinearly binned files. Various routines in binning.py were written to throw an error on nonlinear files, but as far as I can tell there is no reason for this to happen because they function perfectly well, since they only operate on transmitted bins anyway.

emcangi committed:
Correct a comment about npix_per_bin. Bins are not square.

emcangi committed:
Move import of get_wavelengths to where it's used to avoid a circular import error. Ironically, get_wavelengths was moved out of instrument.py and into echelle.py in 1466a7ae445f33fd4f3b508c88aa82f8c71cd9eb because it fits better, but also to ...avoid a circular import error.

emcangi committed:
Add support for making quicklooks for nonlinearly binned files.

planetarymike commented on maven_iuvs/binning.py:
```diff
@@ -143,16 +144,14 @@ def get_pix_range(myfits, which="spatial"):
     pixlo = myfits['Binning'].data[f'{which[:3]}pixlo'][0]
     pixhi = myfits['Binning'].data[f'{which[:3]}pixhi'][0]
 
-    if not (set((pixhi[:-1]+1)-pixlo[1:]) == {0}):
```
The issue there is that spaepixlo and spaepix high include the indices of bins that are not transmitted, so there needs to be a cross-check against spaebintransmit to see whether or not the bin is actually included in the file. I think at some point in some application of this function I worried about dividing the data by compute number of pixels per bin, and that division won't work if the bin widths are calculated using a diff on this function. Maybe it's OK now.

planetarymike replied:
I took a closer look and the issue is in `get_npix_per_bin()` . For nonlinear binning schemes that function needs to be modified to only report transmitted bins. Probably an easy fix but for now the nonlinear warning can be moved there

planetarymike replied:
It also looks like some of the echelle graphics routines will need to be modified to account for this too if the nonlinear exception is removed...

emcangi replied:
When I tested this, it produced expected output, with the only difference being that the untransmitted bins leave a much larger "gap" in the pixels. E.g., for a file that is linearly binned with width 11 pix/bin in transmitted bins, and 77 pix/bin in two untransmitted bins in the ~middle of the detector, just above the slit:

```python
>>> print(f"file: {lf['primary'].header['FILENAME']}")
file: mvn_iuv_l1a_outdisk-orbit02102-ech_20151029T030345_v13_r01.fits
>>> print(f"Binning scheme: {iuvs.binning.get_binning_scheme(lf)}")
Binning scheme: {'bintbl': 'NON LINEAR 4,4 nonlin_0002', 'nspa': 50, 'nspe': 160}
>>> print("Transmitted/untransmitted bins: ", lf['binning'].data['spabintransmit'])
Transmitted/untransmitted bins:  [[0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 1
  1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0]]
>>> print(f"Bin widths: {lf['binning'].data['spabinwidth']}")
Bin widths: [[100 100  99  11  11  11  11  11  11  11  11  11  11  11  11  11  11  11
   11  11  11  11  11  11  11  11  11  11  77  77  11  11  11  11  11  11
   11  11  11  11  11  11  11  11  11  11  11  11  11  11  11  11  11  11
   11  17   4]]
>>> print(f"Lowest pixel number in bin: {lf['binning'].data['spapixlo']}")
Lowest pixel number in bin: [[299 310 321 332 343 354 365 376 387 398 409 420 431 442 453 464 475 486
  497 508 519 530 541 552 563 728 739 750 761 772 783 794 805 816 827 838
  849 860 871 882 893 904 915 926 937 948 959 970 981 992]]
>>> print(f"Highest pixel number in bin: {lf['binning'].data['spapixhi']}")
Highest pixel number in bin: [[ 309  320  331  342  353  364  375  386  397  408  419  430  441  452
   463  474  485  496  507  518  529  540  551  562  573  738  749  760
   771  782  793  804  815  826  837  848  859  870  881  892  903  914
   925  936  947  958  969  980  991 1002]]
>>> print(f"results of get_pix_range(): {iuvs.binning.get_pix_range(lf, which='spatial')}")
results of get_pix_range(): [ 299  310  321  332  343  354  365  376  387  398  409  420  431  442
  453  464  475  486  497  508  519  530  541  552  563  574  739  750
  761  772  783  794  805  816  827  838  849  860  871  882  893  904
  915  926  937  948  959  970  981  992 1003]
```

From what I can tell get_pix_range() is only used in: 
- echelle.py
- echelle_graphics.py
- geometry.py (one call, and it's not even imported, so I think it must not have ever been used?)
- line_fit_plot.py

From what I can tell in my testing, it's working in all of those except geometry.py, but I haven't tested it for use with the standard resolution. Is there another case you're thinking of where this could break?

Also, these commits weren't supposed to be on this branch since they're not strictly related to IPH. I can look more into moving them to another branch later, at a brief glance it looked like it was going to be a bit of a headache.

planetarymike replied:
`get_pix_range()` is called by `get_npix_per_bin`, which is called by `detector_image`. In that routine the following happens (this is similar to some of your echelle routines I think):

```python
def detector_image(myfits, integration=0, **kwargs):
    data = myfits['detector_dark_subtracted'].data[integration]

    spapixrange = get_pix_range(myfits, which="spatial")
    spepixrange = get_pix_range(myfits, which="spectral")
    npixperbin = get_npix_per_bin(myfits)
            
    data = data / npixperbin

    fig = plot_detector(data, spapixrange, spepixrange, **kwargs)
```

If get_pix_range returns extra entries for non-transmitted bins, there will be a dimensional mismatch between the `data` element and `npixperbin`. Additionally, `plot_detector()` will have issues because the `pcolormesh` call will have a dimensional mismatch between the data and the pixel boundaries given. 

If we want to retain the new behavior of `get_pix_range()`, that's OK, but the downstream routines would either need to be fixed to operate correctly with nonlinear binning or raise an error similar to the old error raised by `get_pix_range()` in the event a file with nonlinear binning is encountered.

planetarymike replied:
Maybe I'm still missing something--- if so we should chat in person!

emcangi replied:
Its probably hard to discuss in thread but:

> If get_pix_range returns extra entries for non-transmitted bins, 

I was saying this doesn't happen as far as I can tell, at least when run on echelle files. In my snippet above, in the last line, it jumps from 574 pix to 739 whereas the rest are diff'ed by 11 pix only. I was able to make quicklooks with it (as I showed you in person, and I have pasted it below again to link it to this discussion).


<img width="3296" height="2112" alt="image" src="https://github.com/user-attachments/assets/5ff8260d-44cc-41c2-bada-a0ffa4a783d9" />


I checked every entry in the index file where the binning is nonlinear, and they all have the same scheme, so they all should work (not gonna test it because there's 2400+ files and it will take forever). 

I can still run `detector_image()` / `plot_detector()` successfully on a linearly-binned l1b FUV file. I'm happy to test on a nonlinearly binned l1b FUV file if there were any (thinking maybe not? I went back to the range where echelle was being nonlinearly binned, and those FUV files were linearly binned), but I'm less familiar with it and don't have it downloaded.

I suppose it's possible in the future we could change the binning again, so we should build in a catch? Would that satisfy this thread?

emcangi replied:
Okay, I've added an assertion in the detector image routines here: d6d354e

Let me know if that looks good enough to complete the review!

planetarymike replied:
I just pushed some changes to fix the remaining issues.

Nonlinearly binned files could be handled pretty well by the previous code, but the issue was that one transmitted pixel would be assigned an incorrect dimension that extended over any non-transmitted bins to the lower edge of the next transmitted bin. Now this does not occur--- binning functions have been modified to return values that are correct regardless of linear / nonlinear binning, and plotting functions have been modified to pad data arrays with NaN values for nontransmitted bin ranges.

emcangi replied:
Great!

planetarymike committed:
fixed dynesty fitting when background is crazy (happens rarely).

reparameterized background function so that polynomial coefficients have similar magnitude.

Added quadratic and cubic background term. (Cubic seems to not be justified to my eye, even quadratic could go if needed.)

emcangi committed:
Correct a bug where 'hand-filled' dictionaries were being added to the list of fit param dictionaries BEFORE those generated by the fitter, which would have caused issues on any files with bad frames on any but the last observation.
Add logic to return nan arrays if an observation breaks the fitting algorithm, so that processing the file at large can still proceed.
Set a bound for the residuals axis in the event that a frame is totally broken so that a plot can still be made with nothing on it.
Correct a few comments, remove old arguments from docstrings.

planetarymike committed:
Remove cubic term from background fit

emcangi committed:
Update the fitting of IPH line center to instead be a fit of the displacement of IPH from the fitted H line.
Change name from "central_wavelength_IPH" to "delta_wavelength_IPH" since it is more descriptive.

emcangi committed:
Correct a check for whether a file is nonlinear in the quicklook code

emcangi committed:
Revert "Update the fitting of IPH line center to instead be a fit of the displacement of IPH from the fitted H line." For some reason, adding together "delta_wavelength_IPH" and "central_wavelength_H" in lineshape_model() slows down the fitting significantly (adds 5-10 seconds to every call to convert_l1a_to_l1c).

This reverts commit 6795164877bbaa7fd02e2528f26cfc204b03e771.

emcangi committed:
Adjust plot_line_fit() so that the guidelines of different emission lines are not plotted if the frame is bad, thus removing the requirement to make up fake values for the parameter dictionary in echelle.py.
While we're at it, also make it so the background isn't plotted if the fit was bad.

emcangi committed:
Add back the docstring of predict_IPH_linecenter which was lost in the revert commit earlier.

emcangi committed:
Add an assertion statement to ensure that the dimensions of npixperbin and data match, following on to removal of exception generation in get_npix_per_bin().

emcangi committed:
Correct bug in b91432e5e2fe5b83aee03d9160e029134521defc that would prevent fitting a file at all if it was recently reprocessed, even if writeout was not requested.

emcangi committed:
Fix bug with old variable name when fitting with an alternate background style.

emcangi committed:
Fix typo in parameter names

planetarymike committed:
Modify binning and plot function to properly handle nonlinearly binned files

emcangi committed:
Make the fit_1sigma list in the event of bad frames be based off of the returned list, rather than the list for the parameters. They should nominally be the same, but this is safer.

emcangi committed:
Update code after jax implementation in e6810284245d86a86e0dc5d4877b255d53a3f12d to support the inclusion of the alternate background (currently called "BU_bg"):
- Create a copy of BU_bg array in fit_H_and_D() that is a jax array, for use in sending into objfn_args.
- Allow for the case where BU_bg may be alternately np.nan (required by jax instead of "None"), or an array of non-nan values by applying .all() to most checks of np.isnan() or jnp.isnan().
- Remove the index of the BU_bg argument from the list of static_argnums.
- Update the deletion of entries in the hessian array to first figure out which indices must be removed in either cases where IPH is not fit or BU background IS fit.
- Remove the code that deleted parameters from the initial guess, since all parameters/bounds must be sent to the solver at all times with the jax implementation.
- Gussy up the bounds comment strings.
- Improve make_fit_param_dict to determine the indices to use with a simple length comparison, rather than making a decision based on the value of BU_bg (this function was called in multiple places, making BU_bg sometimes be a jax array and sometimes be a normal numpy array, which caused downstream errors).
Related problems fixed:
- Preclude the max_ll (maximum log likelihood value) from being sent into make_fit_param_dict, as it's not part of the parameter list anyway and is better handled in the functions that exclusively deal with it.
- Clarify the possible exceptions when fitting is don by adding a printed warning.

emcangi committed:
Remove commented-out quadratic term. If we want it in the future we can revert bbdb280f93cc16eca807b9462f658564af2ba95d.

planetarymike committed:
modify get_ech_slit_indices to call pix_to_bin

emcangi reviewed (commented):
LGTM

planetarymike reviewed (approved):
I think this is ready to go!

emcangi commented:
Just confirming - I went back and checked the file where the H line was being severely underfit with the IPH updates. That file is now fine, so I think the issue was related to the lack of geometry - that or other changes we made in the process fixed the problem.

(file was: mvn_iuv_l1a_inlimb-orbit24112-ech_20250801T032926_v13_XYY.fits.gz, where x= r|s and y=\d)
